torch/_dynamo/variables/functions.py

```diff
@@ -492,6 +492,19 @@ def reconstruct(self, codegen):
             codegen(self.fn_name)
         codegen.extend_output([create_instruction("MAKE_FUNCTION", arg=flags)])
 
+        # When graph breaks happen, the global variables of resumed function
+        # come from the outermost function which may not have globals needed
+        # by this nested function, we have to add these missing globals.
+        root_globals = codegen.tx.output.root_globals
+        for key in self.code.value.co_names:
+            if key not in self.f_globals:
+                continue
+            if key in root_globals:
+                if not (key.startswith("__") and key.endswith("__")):
+                    assert root_globals[key] == self.f_globals[key]
+            else:
+                root_globals[key] = self.f_globals[key]
+
```

Comment:
This is modifying the user's global scope, copying every object from one user module to another user module.  This is not correct and could break user code.

The root problem is that f_globals is pointing to the wrong module.  Which I believe is coming from the usage of the `MAKE_FUNCTION` bytecode above.

Instead of using `MAKE_FUNCTION`, I think we can call:
```
types.FunctionType(code, globals, name, argdefs, closure)
```
directly.  The constructor of `FunctionType` allows passing a globals pointer different from the current module.